vql/networking/tls.go

```diff
@@ -0,0 +1,150 @@
+	config_proto "www.velocidex.com/golang/velociraptor/config/proto"
+	"www.velocidex.com/golang/velociraptor/crypto"
+)
+
+// hashCertificate takes a tls.Certificate and return the sha256
+// fingerprint of said certificate. The return value is the hex
+// representation of the byte sequence returned by the hashing
+// function.
+func hashCertificate(cert *x509.Certificate) (string, error) {
+	h := sha256.New()
+
+	if _, err := h.Write(cert.Raw); err != nil {
+		return "", err
+	}
+
+	checksum := h.Sum(nil)
+	return hex.EncodeToString(checksum), nil
+}
+
+// If we deployed Velociraptor using self signed certificates we want
+// to be able to trust our own server. Our own server is signed by our
+// own CA and also may have a different common name (not related to
+// DNS). For example, in self signed mode, the server certificate is
+// signed for VelociraptorServer but may be served over
+// "localhost". Using the default TLS configuration this connection
+// will be rejected.
+
+// Therefore in the special case where the server cert is signed by
+// our own CA, and the Subject name is the pinned server name
+// (VelociraptorServer), we do not need to compare the server's common
+// name with the url.
+
+// This function is based on
+// https://go.dev/src/crypto/tls/handshake_client.go::verifyServerCertificate
+func customVerifyConnection(
+	CA_Pool *x509.CertPool,
+	config_obj *config_proto.ClientConfig) func(conn tls.ConnectionState) error {
+
+	// Check if the cert was signed by the Velociraptor CA
+	private_opts := x509.VerifyOptions{
+		CurrentTime:   time.Now(),
+		Intermediates: x509.NewCertPool(),
+		Roots:         x509.NewCertPool(),
+	}
+	if config_obj != nil {
+		private_opts.Roots.AppendCertsFromPEM([]byte(config_obj.CaCertificate))
+	}
+
+	return func(conn tls.ConnectionState) error {
+		// Used to verify certs using public roots
+		public_opts := x509.VerifyOptions{
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+			DNSName:       conn.ServerName,
+			Roots:         CA_Pool,
+		}
+
+		// First parse all the server certs so we can verify them. The
+		// server presents its main cert first, then any following
+		// intermediates.
+		var server_cert *x509.Certificate
+
+		for i, cert := range conn.PeerCertificates {
+			// First cert is server cert.
+			if i == 0 {
+				server_cert = cert
+
+				// Check the fingerprint of the certificate first. If there
+				// is no match, fall back to testing against our CA cert.
+				if len(config_obj.GetCertificateHashes()) > 0 {
+					certSha256, err := hashCertificate(cert)
+					if err != nil {
+						return err
+					}
+
+					for _, hash := range config_obj.GetCertificateHashes() {
+						if hash == certSha256 {
+							// we found a matching cert, connection can continue
+							return nil
+						}
+					}
+
+					// if certificate pinning is enforced, we need to abort the
+					// connection when there was no match
+					if config_obj.ForceCertificatePinning {
```

Comment:
Maybe uplift this condition so we dont bother with cert hashes unless we need force pinning?

Reply:
I believe this depends on whether or not the idea of being able to short-circuit cert validation may stay (think of the short-circuiting as a means of "if you see this hash, just accept the certificate"). If we did the ForceCertificatePinning check beforehand, short-circuiting wouldn't work anymore because we would skip the hashing when the value is true.

Reply:
Obseleted by using an enum-style approach to the verification mode now.